### DISCUSSION TITLE
ccache-friendly way to embed git sha?

### DISCUSSION THREAD

**Discussion by vadi2** [ccache/ccache#1138] [Questions & answers] [answered]
I embed git sha's in test binaries using `-DAPP_BUILD="-testing-pr<git PR>-<git SHA>"` and since the sha changes on every commit, this throws ccache off.

`CCACHE_IGNOREOPTIONS` could be of help here, however this introduces another problem - the _wrong_ sha is then embedded in the binary as ccache re-uses it.

What's the best way to go about solving this?

**Comment by jrosdahl** [comment]
In this case I don't think there's any way to have your cake and eat it too. What you could do is to pass the `-D` option to only one object file (and maybe let that object file only contain the string with the hash) so that it's quick to recompile anyway.

**Comment by Bezalja** [comment, marked as answer]
Instead of using a compiler define, it's better to use a configured cpp file for this.
Here's a good blogpost on how to do so using CMake:
https://jonathanhamberg.com/post/cmake-embedding-git-hash/
This way, there will be only a single file missing the cache.
